Added new contribute to guides (#501)
AlmaLinux/wiki: docs/Contribute-to-AlmaLinux-Build-System.md

#### A new directory
* Add description to `roles/dev_deploy/defaults/main/common.yml`.

#### A new docker container
* Add a new description of a container's docker image to `roles/dev_deploy/defaults/main/docker_images.yml`.
* Add a new description of a docker container to `roles/dev_deploy/defaults/main/docker_containers.yml`.

## Commit Guidelines

When you are ready to create a pull request with your suggested changes, we'd like to provide some commit guidelines in order to make our commit messages consistent across all repositories around the AlmaLinux Build System. Every contributor could use this document to make commits that:
* Make sense (grammatically).
* Are meaningful/clear enough -> describe the change(s) made to the code.
* Include references to Github issue (whenever possible).
* Include references to fixed (or resolved) issue(s).
* This guide also aims to cover not only atomic commits made by contributors but also how merge commits must be performed.

### Commit messages
For this first iteration, we aim to focus on these main two goals:
* Commit messages must be precise about the changes made so other developers understand what happened in that commit.

> * A person can identify the ticket in Github that motivated such change. If possible link a GitHub issue that the commit(s) fixed or resolved. 

Reviewer: ```suggestion
* A person can identify the issue in Github that motivated such change. If possible link a GitHub issue that the commit(s) fixed or resolved. 
```

* There is only a single commit that fixes a single issue.
* When there is a list of meaningul commits that fixes one or several issues at a time.
* When there's no issue involved.

Even if we don't want to be very pedantic about how commits look like (unless we decide to strengthen the policy for other reasons), we'd like to be more or less consistent with the goals described above.

#### A very detailed commit message

The commit message should meet the requirements:
* Should contain short (72 chars or less) summary.
* Should contain more detailed explanatory text. Wrap it to 72 characters. The blank line separating the summary from the body is critical (unless you omit the body entirely).
* Further paragraphs come after blank lines.
* Bullet points are okay, too.
    * Typically a hyphen or asterisk is used for the bullet, followed by a single space. Use a hanging indent.
* If want to add references to fixed or resolved issue(s), do it at the end as in the exmaple: 
  ```
  Fixes: https://github.com/AlmaLinux/build-system/issues/80.
  ```
  or 
  ```
  Resolves: https://github.com/AlmaLinux/build-system/issues/80.
  ```

> Ideally, every commit message should look like this, but we try to be flexible enough.

Reviewer: I'd love to see us link to any commit message that incorporates all of these guidelines (especially if the example shows a list being used in addition to linking to an existing issue), so folks can see a real-world implementation of this guide.

Reviewer 2: Here's one https://github.com/AlmaLinux/albs-web-server/commit/db062d4e69fe7bc93c59ce5c4c977e3fc3667419

#### Summary

> Perfect summary should meet further requirements: 

Reviewer: ```suggestion
The perfect summary will meet these requirements: 
```

* Link a fixed/resolved issue:
  ```
  Fixes: https://github.com/AlmaLinux/build-system/issues/80
  ```
  or
  ```
  Resolves: AlmaLinux/build-system/issues/80
  ```
  It's optional, and can be used only when the commit message is short enough to fit in 72 characters.

#### Body

Body is a more detailed explanatory text. Wrap it to 72 characters. The blank line separating the summary from the body is critical (unless you omit the body entirely).

Further paragraphs come after blank lines.

* Bullet points are okay, too.
* Typically a hyphen or asterisk is used for the bullet, followed by a single space. Use a hanging indent.

When writing the body part, you can add as much details as you want on the commit message. Just remember:
* The body is optional if the subject is enough to describe the commit.
* Wrap lines to 72 characters.
* Using the imperative or simple past tenses is acceptable, but please be consistent.
* Feel free to describe what you think it's worth being described as if you were your colleague (that is more or less familiar to the code) that would like to understand what happened by just reading a few lines.

#### Footer

> Footer is a perfect place to add references as links to fixed or resolved issue(s). If have not referenced issues in the subject, then they can be added here:

Reviewer: ```suggestion
The footer is a perfect place to add references as links to fixed or resolved issue(s). If you have not referenced issues in the subject, then they can be added here:
```

```
Fixes https://github.com/AlmaLinux/build-system/issues/80
```
or 
```
Resolves https://github.com/AlmaLinux/build-system/issues/80
```

Remember:
* Not having referenced issues is a possibility in case of dependency version updates and so on.
* Can use a bullet list for the list of references.

> * Although using Github's abbreviations is okay, we privilege using the full URL as it will be meaningful if somebody is not reading the commit message through Github's interface.

Reviewer: ```suggestion
* Although using Github's abbreviations is okay, we require using the full URL as it will be meaningful if somebody is not reading the commit message through Github's interface.
```

### Everyday Use cases

#### The simplest commit message

```
Fixed create_module_by_payload (#123)
```
As you can see, simple commit messages like this one are ok. Even if not providing a good description in the commit message, it:

* Describes that there was a fix in `create_module_by_payload`.
* References a github issue (#123).

However, this structure might not fit as `#123` will refer to issues/pull request within the same repository. That reference needs to be more specific.

#### A tad more complex one

```
Added some missing AlmaLinux beta repositories

We are adding missing AlmaLinux 8 RT repos for s390x and ppc64le
Also, adding aarch64 NFV repos for AlmaLinux 9

Resolves: https://github.com/AlmaLinux/build-system/issues/80
```

> 

Reviewer: ```suggestion
In this structure you can see that we have used an expanded example that:

* adds more context for the changes that were made
* includes the full URL for which issue was fixed
```

#### A little bit about merge commits

> Some commits might not be meaningul, others might not include references, some can ollow the same structure or are not fully descriptive. Since 99% of the time we merge pull requests (hereinafter PRs), we should add the relevant data in the mere commit message. By relevant data we understand the branch the code is coming from and the information we would like to know about the commits as described earlier in the document.

Reviewer: ```suggestion
Some commits might not be meaningful or might not include references to specific GitHub issues, so they cannot follow the same structure or be as fully descriptive. Since 99% of the time we merge pull requests (hereinafter PRs), we should add the relevant data in the commit message. By relevant data we mean the branch the code is coming from, and the information we would like to know about the commits as described earlier in the document, so that it can be understood later.
```
I'm not entirely sure I understood what this paragraph was trying to say, so if my suggestions weren't good, please feel free to say so.

#### Single commit PRs

> Sometimes, a PR has a single and meaningful commit message that doesn't require to reference an issue, i.e.:

Reviewer: ```suggestion
Sometimes, a PR has a single and meaningful commit message that doesn't require referencing an issue, i.e.:
```

`Bump Pydantic from 1.10.7 to 1.10.8`

In this case, we can do a squash and merge for the commit to end up in the repository in the following way:

`Bump Pydantic from 1.10.7 to 1.10.8 (#562)`

You can also create a merge commit message this way.
```
Merge pull request #562 from javihernandez:pydantic-1.10.8

Bump Pydantic from 1.10.7 to 1.10.8
```

Of course, and as always, ensure that if the change is motivated by a Github issue, please do add a reference in the commit message as described in the previous section of the document.

#### Several commits PRs
Usually, PRs involve more than one single commit. Ideally, they all should follow the structure described through this document, but sometimes this is not the case. We can see commit messages like:

> * Addressed review comments.
> * Fixed typo.
> * etc.
> Which, for obvious reasons, do not provide a lot of information about the commit. For this reason, all our merge commit messages must:
> 
> * Reference the branch being merged.
> * Reference the pull request.

Reviewer: ```suggestion
* Addressed review comments.
* Fixed typo.
* etc.
Which, for obvious reasons, do not provide a lot of information about the commit. For this reason, all our merge commit messages must:

* Reference the branch being merged.
* Reference the pull request.
```
Can you add examples of the erroneous examples being fixed here?  Something like "* Addressed review comments for $branch" or something? Based on the description, I'm not sure how those commit messages should specifically be formatted.

---
title: 'AlmaLinux Build System'
---

###### last updated: 2024-09-20

# Contribute to AlmaLinux Build System

The AlmaLinux Build System is managed by the [SIG/Build System](/sigs/Build-System) SIG. 

The project is open for community contributions. If you are interested in contributing to the Build System project, we recommend you start by checking the AlmaLinux Build System project [milestones](https://github.com/AlmaLinux/build-system/milestones) for ongoing tasks and following the guidelines below.

When you are ready to create a pull request with your suggested changes, please follow the **[commit guidelines](#commit-guidelines)** to keep our commit messages consistent across all repositories around the AlmaLinux Build System. 

## Contribute to the codebase

To contribute to development we recommend deploying the AlmaLinux Build System [locally](https://github.com/AlmaLinux/build-system/wiki/Deploy-on-local-and-remote-machines) and following these guide steps for possible modifications and commit guidelines.

> ### Help Wanted

Reviewer 2: Hey @sboldyreva, I actually talked to the team and we compiled the following list:
* Help on keeping our Build System SIG documentation up to date (Readmes in repos, wiki, docs, sig wiki page, etc)
* Testing (experience with pytest), we need help to:
  * Increase our test coverage in repos that already have tests
  * Add tests to those that don't have tests at all
  * Design and implement of integration/e2e tests that involve different services
* Famliar with Ansible? Help us in testing and improving our current ansible roles to deploy ALBS
* Interested in supply-chain security and SBOM? Help us in defining next steps towards providing and expanding the current SBOM data that AlmaLinux is generating

Also, I think that we can provide a short list of our tech stack, something like:
* Backend: Python, FastAPI, SQLAlchemy, PostgreSQL, redis
* Frontend: JavaScript, Vue.js, Quasar
* Tooling: Docker, Docker Compose, Ansible, Terraform

What do you think?

Reviewer: Love this description here.

Author: @bennyvasquez @javihernandez added

We are seeking contributors to help us with:
* Python and JavaScript Developers are needed to improve UI/UX.
* Add the ability to run test VMs in different clouds (AWS, Azure, etc.).
* Add the ability to use external repositories for testing scenarios (e.g. LTP for kernel, CentOS tests for different packages, openQA, etc.).
* Add OpenStack backend driver support.
* Add Azure backend driver support.
* Improve ability to delete a build.
* Add Kubevirt backend support.

> #### A new service's config

Reviewer 2: I'm not sure if it makes sense to have this and next 2 sections.

In addition to the `help wanted` section, do you think that we can add a list of our git repos?
Yes, I know that they are already in [build-system repo](https://github.com/AlmaLinux/build-system?tab=readme-ov-file#development), but I guess that we can add the most relevant ones here, let's say web-server, frontend, node, sign-node, albs-deploy and alts

Author: These sections have been there from the very first version of contributions guidelines, if you confirm they are not relevant and needed anymore, I'm okay with removing these :) 

As for the help wanted part - Alexandra consulted the build system team and helped me update the items. We also have the fresh repos list on the Build System SIG page, but I'd prefer having them listed along with the updated structure scheme. Would you like to help me with it?

Author: That was the plan to add that suggestion, but I was also interested in this section. Thanks for confirming this, I'll delete the info :)

* Add template to `roles/dev_deploy/templates` for config of a service. The name of the file should be `<name_of_target_config_file>.j2`.
* Add description of config to `roles/dev_deploy/defaults/main/configs.yml`.

#### A new directory
* Add description to `roles/dev_deploy/defaults/main/common.yml`.

#### A new docker container
* Add a new description of a container's docker image to `roles/dev_deploy/defaults/main/docker_images.yml`.
* Add a new description of a docker container to `roles/dev_deploy/defaults/main/docker_containers.yml`.

> ## Commit Guidelines

Reviewer 2: Actually, I'm not sure we should add the full commit guidelines here. What if we keep it as a standalone page? What do you think?

Reviewer: I think keeping them in the contributor's guide is the right move, since each part of the AlmaLinux project might require different things.

Reviewer 2: I don't mind keeping them here, it's just that, even though the commit guidelines are important, I don't really believe they deserve to be half of the contribution guide. That's my reasoning behind, but no strong opinion.

Author: From the other side, folks can notice them easier that also leads to their commits being up to your expectations

Reviewer 2: Good point @sboldyreva, I totally agree, let's just keep them here
